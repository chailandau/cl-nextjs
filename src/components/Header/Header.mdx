import { ArgTypes, Canvas, Meta } from '@storybook/blocks';

import * as HeaderStories from './Header.stories';

<Meta of={HeaderStories} />

# Header Component

The `Header` component is on every page. It contains the logo along with the main navigation (using the [Menu](?path=/docs/molecules-menu--docs) molecule).

<Canvas of={HeaderStories.Default} />

## menuItems

List of navigation links to display.

<br />
<br />
<br />

### Props Table

<ArgTypes />
