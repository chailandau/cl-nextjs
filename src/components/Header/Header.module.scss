.header {
    &,
    &:before {
        @include borderRadius;
    }
    position: relative;
    margin: rem(10) rem(10) 0 rem(10);
    gap: rem(40);
    justify-content: space-between;
    align-items: center;
    @include boxShadow;

    .logo,
    .right-content {
        z-index: 101;
    }
    .logo {
        width: rem(44);
        padding: unset;
        @include hover {
            @include hoverOpacity;
            & > span {
                &:before {
                    display: none !important;
                }
            }
        }
    }
}
