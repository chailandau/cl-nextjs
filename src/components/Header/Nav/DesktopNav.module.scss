.desktop-nav {
    display: none;
    @include tabletLg {
        display: block;
    }
}
