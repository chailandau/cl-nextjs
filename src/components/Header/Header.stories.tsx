import type { Meta, StoryObj } from '@storybook/react';

import Header from './Header';

import { menuItems } from '@/__mocks__/Header.mock';

const meta: Meta<typeof Header> = {
    title: 'Components/Header',
    component: Header,
    argTypes: {
        icon: {
            table: {
                disable: true
            }
        }
    }
};

export default meta;

type Story = StoryObj<typeof Header>;

export const Default: Story = {
    args: {
        menuItems
    }
};
