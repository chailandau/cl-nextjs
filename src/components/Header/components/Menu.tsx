import { FC } from 'react';

import { HeaderProps } from '../Header';

import styles from './Menu.module.scss';

import Link from '@/atoms/Link/Link';
import { getPageLabel } from '@/utils/getPageLabel';

export interface MenuProps extends HeaderProps {
    /** Optional classname */
    className?: string;
}

const Menu: FC<MenuProps> = ({ menuItems, className = styles['menu'] }) => {
    const menuContent = menuItems?.map((menuItem) => {
        if (!menuItem.page) {
            return null;
        }

        const menuLabel = getPageLabel(menuItem);

        return (
            <li key={menuItem?.page?.id}>
                <Link
                    key={menuItem?.page?.id}
                    href={`${process.env.NEXT_PUBLIC_BASE_URL}/${menuItem?.page?.slug}`}
                    className={styles['menu-link']}
                >
                    {menuLabel}
                </Link>
            </li>
        );
    });

    return <ul className={className}>{menuContent}</ul>;
};

export default Menu;
