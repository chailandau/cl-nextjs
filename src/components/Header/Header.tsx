'use client';

import FocusTrap from 'focus-trap-react';
import { FC, useEffect } from 'react';

import styles from './Header.module.scss';

import { Nav_MenuItems } from '@/api/graphqlTypes';
import Logo from '@/assets/svg/Logo.svg';
import Image from '@/atoms/Image';
import Link from '@/atoms/Link';
import { DesktopNav, MobileNav } from '@/components/Nav';
import Flex from '@/molecules/Flex';
import { MenuToggle } from '@/molecules/Menu';
import Section from '@/molecules/Section';
import useStore from '@/store/useStore';
import { laptopQuery, useMediaQuery } from '@/utils/hooks/useMediaQuery';

export interface HeaderProps {
    /* Nav to pass into header */
    menuItems?: Nav_MenuItems[] | null;
}
const Header: FC<HeaderProps> = ({ menuItems }) => {
    const isLaptop = useMediaQuery(laptopQuery);

    const { menuOpen, setMenuOpen } = useStore();

    useEffect(() => {
        if (isLaptop) {
            setMenuOpen(false);
        }
    }, [isLaptop]);

    return (
        <FocusTrap active={menuOpen}>
            <Section as='header' className={styles['header']}>
                <Link
                    href={process.env.NEXT_PUBLIC_BASE_URL as string}
                    className={styles['logo']}
                    underline={false}
                >
                    <Image src={Logo} alt='C.' priority hasBorder={false} />
                </Link>

                <Flex className={styles['right-content']}>
                    {isLaptop && <DesktopNav menuItems={menuItems} />}

                    <MenuToggle />
                </Flex>
                {!isLaptop && menuOpen && <MobileNav menuItems={menuItems} />}
            </Section>
        </FocusTrap>
    );
};

export default Header;
