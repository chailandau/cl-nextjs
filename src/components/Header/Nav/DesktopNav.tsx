import { FC } from 'react';

import styles from './DesktopNav.module.scss';

import { HeaderProps } from '@/components/Header';
import Flex from '@/molecules/Flex';
import { Menu } from '@/molecules/Menu';
import LazyMotionDom from '@/utils/framer/LazyMotionDom';

const DesktopNav: FC<HeaderProps> = ({ menuItems }) => (
    <LazyMotionDom>
        <Flex as='nav' className={styles['desktop-nav']}>
            <Menu menuItems={menuItems} icon={false} />
        </Flex>
    </LazyMotionDom>
);

export default DesktopNav;
