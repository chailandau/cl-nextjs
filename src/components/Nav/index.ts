import DesktopNav from './components/DesktopNav';
import MobileNav from './components/MobileNav';

export { DesktopNav, MobileNav };
