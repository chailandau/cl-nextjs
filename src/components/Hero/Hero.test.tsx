import { composeStory } from '@storybook/react';
import { act, fireEvent, render, screen } from '@testing-library/react';

import HeaderMeta, { Default as HeaderDefault } from './Hero.stories';

import { menuItems, menuItemsNullPage } from '@/__mocks__/Header.mock';
import { mockIsTabletLg } from '@/__mocks__/useMediaQuery.mock';
import { DesktopNav, MobileNav } from '@/components/Header/Nav';
import { Menu, MenuToggle } from '@/molecules/Menu';
import { testAxeAndSnapshot } from '@/utils/testHelpers';

const Header = composeStory(HeaderDefault, HeaderMeta);

describe('Header', () => {
    describe('Menu', () => {
        it('renders menu items correctly', () => {
            render(<Menu menuItems={menuItems} />);
            const menuItemsElements = screen.getAllByRole('listitem');
            expect(menuItemsElements).toHaveLength(3);
        });
        it('returns null if page in menuItems is null', () => {
            render(<Menu menuItems={menuItemsNullPage} />);
            expect(screen.queryByRole('listitem')).toBeNull();
        });
        testAxeAndSnapshot({ component: <Menu menuItems={menuItems} /> });
    });
    describe('Menu Toggle', () => {
        it('opens navigation on click', async () => {
            const header = render(<Header />);
            const menuToggle = await header.findByLabelText('menu toggle');
            fireEvent.click(menuToggle);
            expect(menuToggle).toHaveClass('open');
        });
        testAxeAndSnapshot({ component: <MenuToggle /> });
    });
    describe('Mobile Nav', () => {
        beforeEach(async () => {
            act(() => {
                mockIsTabletLg(false);
            });
            render(<Header />);
            fireEvent.click(await screen.findByLabelText('menu toggle'));
        });
        it('removes mobile nav when isTabletLg is true', async () => {
            act(() => {
                mockIsTabletLg(true);
            });
            expect(await screen.findByRole('navigation')).not.toHaveClass(
                'mobile-nav'
            );
        });
        testAxeAndSnapshot({ component: <MobileNav menuItems={menuItems} /> });
    });
    describe('DesktopNav', () => {
        const DesktopNavEl = <DesktopNav menuItems={menuItems} />;
        beforeEach(() => {
            render(DesktopNavEl);
            act(() => {
                mockIsTabletLg(true);
            });
        });
        it('renders correctly', () => {
            expect(screen.getByRole('navigation')).toBeInTheDocument();
        });
        testAxeAndSnapshot({ component: DesktopNavEl });
    });
});
