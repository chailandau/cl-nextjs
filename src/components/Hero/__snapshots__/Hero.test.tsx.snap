// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hero matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="flex section hero"
  >
    <div
      class="flex icon"
    >
      <div
        class="flex container"
      >
        <img
          alt="Waving hand emoji"
          data-nimg="1"
          decoding="async"
          height="64"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fwave.png&w=128&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fwave.png&w=64&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fwave.png&w=128&q=75 2x"
          style="color: transparent;"
          width="64"
        />
      </div>
    </div>
    <h1
      class="heading lg primary"
    >
      Hi, I’m  Chai.
    </h1>
    <div
      class="colored-subhead colored-subhead"
    >
      <span
        class="text md secondary"
      >
        I’m a 
      </span>
      <span
        class="text md bg-aqua secondary"
      >
        designer-turned-engineer
      </span>
      <span
        class="text md secondary"
      >
         obsessed with 
      </span>
      <span
        class="text md bg-white secondary"
      >
        accessibility
      </span>
      <span
        class="text md secondary"
      >
         and 
      </span>
      <span
        class="text md bg-pink secondary"
      >
        user experience.
      </span>
    </div>
    <a
      class="link underline cta"
      href="/about"
      rel=""
      target=""
    >
      <span
        class="link-text"
      >
        learn more
      </span>
      <span
        class="link-icon"
      >
        &gt;
      </span>
    </a>
  </section>
</DocumentFragment>
`;
